engine/access/ingestion/collection_syncer.go

```diff
@@ -0,0 +1,398 @@
+		}
+		s.requester.EntityByID(collection.ID(), filter.HasNodeID[flow.Identity](guarantors...))
+	}
+
+	if immediately {
+		s.requester.Force()
+	}
+}
+
+// updateLastFullBlockHeight updates the next highest block height where all previous collections have been indexed.
+// No errors expected during normal operation.
+func (s *CollectionSyncer) updateLastFullBlockHeight() error {
+	lastFullBlockHeight := s.lastFullBlockHeight.Value()
+	lastFinalizedBlock, err := s.state.Final().Head()
+	if err != nil {
+		return fmt.Errorf("failed to get finalized block: %w", err)
+	}
+
+	// track the latest contiguous full height
+	newLastFullBlockHeight, err := s.findLowestBlockHeightWithMissingCollections(lastFullBlockHeight, lastFinalizedBlock.Height)
+	if err != nil {
+		return fmt.Errorf("failed to find last full block height: %w", err)
+	}
+
+	// if more contiguous blocks are now complete, update db
+	if newLastFullBlockHeight > lastFullBlockHeight {
+		err := s.lastFullBlockHeight.Set(newLastFullBlockHeight)
+		if err != nil {
+			return fmt.Errorf("failed to update last full block height: %w", err)
+		}
+
+		s.collectionExecutedMetric.UpdateLastFullBlockHeight(newLastFullBlockHeight)
+
+		s.logger.Debug().
+			Uint64("last_full_block_height", newLastFullBlockHeight).
+			Msg("updated last full block height counter")
+	}
+
+	return nil
+}
+
+// findLowestBlockHeightWithMissingCollections finds the next block height with missing collections,
+// returning the latest contiguous height where all collections are present.
+//
+// No errors expected during normal operations.
+func (s *CollectionSyncer) findLowestBlockHeightWithMissingCollections(
+	lastKnownFullBlockHeight uint64,
+	finalizedBlockHeight uint64,
+) (uint64, error) {
+	newLastFullBlockHeight := lastKnownFullBlockHeight
+
+	for currBlockHeight := lastKnownFullBlockHeight + 1; currBlockHeight <= finalizedBlockHeight; currBlockHeight++ {
+		missingCollections, err := s.findMissingCollectionsAtHeight(currBlockHeight)
+		if err != nil {
+			return 0, ErrMissingCollection
+		}
+
+		// return when we find the first block with missing collections
+		if len(missingCollections) > 0 {
+			return newLastFullBlockHeight, nil
+		}
+
+		newLastFullBlockHeight = currBlockHeight
+	}
+
+	return newLastFullBlockHeight, nil
+}
+
+// OnCollectionDownloaded indexes and persists a downloaded collection.
+// This is a callback intended to be used with the requester engine.
+func (s *CollectionSyncer) OnCollectionDownloaded(_ flow.Identifier, entity flow.Entity) {
+	collection, ok := entity.(*flow.Collection)
+	if !ok {
+		s.logger.Error().Msgf("invalid entity type (%T)", entity)
+		return
+	}
+
+	err := indexer.IndexCollection(collection, s.collections, s.transactions, s.logger, s.collectionExecutedMetric)
+	if err != nil {
+		s.logger.Error().Err(err).Msg("could not index collection after it has been downloaded")
+		return
+	}
+}
```

Comment:
This is copied from the original, so no need to make any changes in this PR. I have opened an issue to improve error handling here:
https://github.com/onflow/flow-go/issues/7510